src/include/detail/graph/greedy_search.h

```diff
@@ -0,0 +1,550 @@
+/**
+ * @file   greedy_search.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *
+ */
+
+#ifndef TILEDB_GREEDY_SEARCH_H
+#define TILEDB_GREEDY_SEARCH_H
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <ranges>
+#include <unordered_set>
+#include <vector>
+
+#include "scoring.h"
+#include "utils/fixed_min_heap.h"
+
+static bool noisy = false;
+[[maybe_unused]] static void set_noisy(bool b) {
+  noisy = b;
+}
+
+template </* SearchPath SP, */ class Distance = sum_of_squares_distance>
+auto greedy_search_O2(
```

Comment:
Would be good to give these descriptive names which explain the differences.

Reply:
O0, O1, O2 were different approaches for accelerating greedy_search.  `best_first` has superseded `greedy_search` so we may just want to deprecate / remove `greedy_search` altogether.